### ISSUE TITLE
Cannot subtract Numeric from BigDecimal

### ISSUE THREAD

**Issue by maxfurman** [sorbet/sorbet#3453]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Adef%20main%0A%20%20BigDecimal(2)%20-%20T.let(2%2C%20Numeric)%20%3D%3D%200%0Aend)

```ruby
BigDecimal(2) - T.let(2, Numeric) == 0
```

#### Observed output

```
editor.rb:4: Expected Integer but found Numeric for argument arg0 https://srb.help/7002
     4 |  BigDecimal(2) - T.let(2, Numeric) == 0
                          ^^^^^^^^^^^^^^^^^
    https://github.com/sorbet/sorbet/tree/master/rbi/stdlib/bigdecimal.rbi#L558: Method BigDecimal#- has specified arg0 as Integer
     558 |        arg0: Integer,
                  ^^^^
  Got Numeric originating from:
    editor.rb:4:
     4 |  BigDecimal(2) - T.let(2, Numeric) == 0
                          ^^^^^^^^^^^^^^^^^
Errors: 1
```

#### Expected behavior

Should be able to subtract (or add, etc.) any compatible type or union of those types to/from a BigDecimal.

- - -

The same error is seen with, for example `T.any(Integer, Float)` instead of Numeric. Anything other than a concrete type of number is rejected by the type checker.

**Comment by sharpobject** [comment]
I cannot speak for the actual behavior of BigDecimal `-`, but it seems like [the current RBI](https://github.com/sorbet/sorbet/blob/master/rbi/stdlib/bigdecimal.rbi#L556) thinks that it is not defined on subtypes of Numeric other than the built-in ones. I'll have to try it out.

About `T.any(Integer, Float)`, it would definitely be nice if when BigDecimal defines `-` for both `Integer` and `Float` arguments giving `BigDecimal` then we could treat that as working on `T.any(Integer, Float)` giving `BigDecimal`.

**Comment by sharpobject** [comment]
For the `Tally` class on https://docs.ruby-lang.org/en/2.7.0/Numeric.html, `BigDecimal(2) - Tally.new('||')` gives a zero `Tally`. If we want to assume that subclasses of `Numeric properly implement `coerce`, we can have an additional signature for `-` like so:

```ruby
  sig do
    params(
        arg0: Numeric,
    )
    .returns(Numeric)
```

**Comment by sharpobject** [comment]
Well, I take that back. It causes code just like code you posted, but with `T.any(Integer, Float)` instead of `Numeric` to typecheck, but probably not the way you wanted it to:

```
$ bazel-bin/main/sorbet issue_3453.rb
issue_3453.rb:6: Revealed type: Numeric https://srb.help/7014
     6 |  T.reveal_type(x)
          ^^^^^^^^^^^^^^^^
  From:
    issue_3453.rb:5:
     5 |  x = BigDecimal(2) - T.let(2, T.any(Integer, Float))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 1
```

It's Numeric here but it should really be BigDecimal.

**Comment by maxfurman** [comment]
This is interesting behavior - for number types that BigDecimal understands, `-` returns another BigDecimal, but for other numeric types it seems to return the same type as was passed in.

The simplest change here is probably to modify the BigDecimal rbi to combine the individual signatures for Integer, Float, etc., into one declaration of T.any(Integer, etc.). I can change my own code to add something like `Number = T.type_alias { T.any(Integer, Float, etc.) }` and use that in place of Numeric. I don't think Numeric is used directly very often; I'm only using it in my type signatures, and just to mean "any number"

A more complicated solution is for Sorbet to resolve union types across multiple signatures. Like for example
```ruby
sig { params(arg0: T.any(Integer, String)).returns(Integer) }
sig { params(arg0: Float).returns(Float) }
def foo; end;

x = foo(T.let(2, T.any(Integer, Float)))
T.reveal_type(x) # => T.any(Integer, Float)
```
In other words, if an argument doesn't match any of the signatures, but it _does_ match a union of all the signatures, then the return type is a union of all the return types.

**Comment by sharpobject** [comment]
> In other words, if an argument doesn't match any of the signatures, but it does match a union of all the signatures, then the return type is a union of all the return types.

I would like this.
